### PR TITLE
[IOTDB-3164] Manage the memory of cross space compaction in write memory controller

### PR DESCRIPTION
**Pull request by THUMarkLau**
See [IOTDB-3164](https://issues.apache.org/jira/browse/IOTDB-3164).

This PR is one part of the compaction memory management. It does the following jobs:
- Allocate memory before cross space compaction, and release the memory after the execution.
- Remove the `RewriteCompactionFileSelector`, which is deduplicated with `RewriteCrossSpaceCompactionSelector`.

To complete the compaction memory management, we need to complete the following jobs:
- Manage the memory usage of unsequence inner space compaction, which uses read point performer.
- Design and implement the memory estimator for read chunk performer.
- Manage the memory usage of sequence inner space comapction, which uses read chunk performer.

### CODE REVIEW

**Comment by THUMarkLau** [comment]
I use IoTDB benchmark tests changed after the performance, and the master (commit: a8f757d6e4fecc2251d2042a025960bf29c6e0f3) are compared. The result of the comparison is shown in the figure below. In terms of average performance, the write performance after the change decreases by about 2.8%.

**master branch**
e173a42189cf91925f9f73475afac2b

**with memory management for cross space compaction**
dd9c26571d2659edfb92e40e6d87df9

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/RewriteCrossSpaceCompactionSelector.java [hunk 1]
```diff
@@ -56,6 +75,223 @@ public RewriteCrossSpaceCompactionSelector(
     this.dataRegionId = dataRegionId;
     this.timePartition = timePartition;
     this.tsFileManager = tsFileManager;
+    this.memoryBudget = config.getCrossCompactionMemoryBudget();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/RewriteCrossSpaceCompactionSelector.java:78]: apply this memory from the storage engine, not from the whole system

**Review by qiaojialin — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/RewriteCrossSpaceCompactionSelector.java (hunk 1 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/RewriteCrossSpaceCompactionSelector.java]: reslove

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 2]
```diff
@@ -623,7 +629,7 @@ public class IoTDBConfig {
   private TSEncoding defaultTextEncoding = TSEncoding.PLAIN;
 
   /** How much memory (in byte) can be used by a single merge task. */
-  private long crossCompactionMemoryBudget = (long) (Runtime.getRuntime().maxMemory() * 0.1);
+  private long crossCompactionMemoryBudget = (long) (allocateMemoryForWrite * 0.1D);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:632]:
```suggestion
  private long compactionMemoryBudget = (long) (allocateMemoryForStorageEngine * compactionProportion);
```

**Comment by qiaojialin** [comment]
https://apache-iotdb.feishu.cn/docx/doxcnuh5hy6ee9Qb7uLCYXkFDdc

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossSpaceCompactionTask.java [hunk 3]
```diff
@@ -79,11 +83,14 @@ public CrossSpaceCompactionTask(
     this.unseqTsFileResourceList = tsFileManager.getUnsequenceListByTimePartition(timePartition);
     this.performer = performer;
     this.hashCode = this.toString().hashCode();
+    this.memoryCost = memoryCost;
   }
 
   @Override
   protected void doCompaction() {
     try {
+      SystemInfo.getInstance().addCompactionMemoryCost(memoryCost);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossSpaceCompactionTask.java:92]: what if the allocateMemory throw interrupt exception without succeeding in adding?

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java (hunk 2 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java]: This param is removed

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossSpaceCompactionTask.java (hunk 3 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossSpaceCompactionTask.java:91]: I move allocateMemory in a seperate try-catch block.

File: server/src/main/java/org/apache/iotdb/db/rescon/SystemInfo.java [hunk 4]
```diff
@@ -170,6 +176,27 @@ public synchronized void resetFlushingMemTableCost(long flushingMemTableCost) {
     this.flushingMemTablesCost -= flushingMemTableCost;
   }
 
+  public void addCompactionMemoryCost(long memoryCost) throws InterruptedException {
+    long originSize = this.compactionMemoryCost.get();
+    while (originSize + memoryCost > memorySizeForCompaction
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/rescon/SystemInfo.java:181]: If there are two tasks call this method concurrently, 
Task1 originSize = 10;
Task2 originSize = 10;
Task1.compactionMemoryCost.compacreAndSet(10, 10+30) = 40;
Task2 compareAndSet is always false, and Task2 sleeps forever.

**Review by qiaojialin — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/rescon/SystemInfo.java (hunk 4 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/rescon/SystemInfo.java:181]: The originSize is updated in the while block.

**Review by qiaojialin — APPROVED** [review]

**Review by HTHou — APPROVED** [review]
LGTM

**Pull request merged**